include/supercluster.hpp

```diff
@@ -286,41 +288,63 @@ class Supercluster {
-
-                    b.visited = true;
-                    b.parent_id = id;
-
-                    // accumulate coordinates for calculating weighted center
-                    weight += b.pos * double(b.num_points);
-                    num_points += b.num_points;
-
-                    if (options_.reduce && b.properties) {
-                        // apply reduce function to update clusterProperites
-                        options_.reduce(clusterProperties, *b.properties);
+                    if (!b.visited) {
+                        num_points += b.num_points;
                     }
                 });
 
-                if (num_points == 1) {
-                    clusters.emplace_back(weight, 1, p.id, clusterProperties);
-                } else {
+                auto clusterProperties = p.properties ? *p.properties : property_map{};
+                if (num_points >= options_.minPoints) { // enough points to form a cluster
+                    point<double> weight = p.pos * double(num_points_origin);
+                    std::uint32_t id = static_cast<std::uint32_t>((i << 5) + (zoom + 1));
+
+                    // find all nearby points
+                    previous.tree.within(p.pos.x, p.pos.y, r, [&](const auto &neighbor_id) {
+                        assert(neighbor_id < cluster_size);
+                        auto &b = previous.clusters[neighbor_id];
+
+                        // filter out neighbors that are already processed
+                        if (b.visited) {
+                            return;
+                        }
+
+                        b.visited = true;
+                        b.parent_id = id;
+
+                        // accumulate coordinates for calculating weighted center
+                        weight += b.pos * double(b.num_points);
+
+                        if (options_.reduce && b.properties) {
+                            // apply reduce function to update clusterProperites
+                            options_.reduce(clusterProperties, *b.properties);
+                        }
+                    });
                     p.parent_id = id;
                     clusters.emplace_back(weight / double(num_points), num_points, id,
                                           clusterProperties);
+                } else {
+                    clusters.emplace_back(p.pos, 1, p.id, clusterProperties);
+                    if (num_points > 1) {
+                        previous.tree.within(p.pos.x, p.pos.y, r, [&](const auto &neighbor_id) {
+                            assert(neighbor_id < cluster_size);
+                            auto &b = previous.clusters[neighbor_id];
+                            // filter out neighbors that are already processed
+                            if (b.visited) {
+                                return;
+                            }
+                            b.visited = true;
+                            clusters.emplace_back(b.pos, 1, b.id,
```

Comment:
Probably, `b.properties` can be shared between clusters? Could we change `std::unique_ptr<property_map> properties{ nullptr }` to `std::shared_ptr<const property_map> properties{ nullptr };` and then try sharing properties between clusters if property map is not changed?